build_configuration/configuration.py

```diff
@@ -0,0 +1,48 @@
+#
+# Copyright (c) 2024 Nordic Semiconductor ASA
+#
+# SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+#
+"""build_configuration module to create and parse build configuration."""
+
+import re
+
+
+class BuildConfiguration(dict):
+    """Represents a build system configuration, providing access to KConfig values.
+
+    This class reads configuration data from a specified file and parses it.
+    Configuration data is accessible as a dictionary.
+    """
+
+    config_value_pattern = re.compile(r"(?P<kconfig_name>[A-Za-z0-9_]+)=(?P<kconfig_value>.*)")
+
+    def __init__(self, input_file: str = ".config") -> None:
+        """Initialize a BuildConfiguration object."""
+        super().__init__()
+        try:
+            with open(input_file, "r") as fh:
+                self._config_data = fh.readlines()
+        except FileNotFoundError as e:
+            raise SystemExit(e)
+        self._parse()
+
+    def _parse(self) -> None:
+        """Parse input .config file and populate the configuration dictionary."""
+        for config_line in self._config_data:
+            if re_result := self.config_value_pattern.match(config_line):
+                kconfig_name = re_result.group("kconfig_name")
+                kconfig_value = re_result.group("kconfig_value")
+                if kconfig_value == "y":
+                    # boolean value
+                    kconfig_value = True
+                elif kconfig_value.startswith("0x"):
+                    # hexadecimal value
+                    kconfig_value = int(kconfig_value, base=16)
+                elif kconfig_value.startswith('"') and kconfig_value.endswith('"'):
+                    # string value
+                    kconfig_value = kconfig_value[1:-1]
+                elif kconfig_value.isdecimal():
+                    # int value
+                    kconfig_value = int(kconfig_value, base=10)
```

Comment:
Is it possible that `kconfig_value` or `kconfig_name` is `None`?
I'm wondering if we need to add a generic `else` here and raise an exception.

Reply:
It seems that for `kconfig_name` and `kconfig_value` is not possible to have `None` values although for `konfig_value` there is a possibility to have an empty value or other values which won't be converted under `if_else` statement - in all mentioned cases `kconfig_value` should be stored as it is due to assignments in line 8: `kconfig_value = re_result.group("kconfig_value")', for example:
```
import re
def _test_code(config_line):
    print(f"{config_line=}")
    config_value_pattern = re.compile(r"(?P<kconfig_name>[A-Za-z0-9_]+)=(?P<kconfig_value>.*)")
    if re_result := config_value_pattern.match(config_line):
        kconfig_name = re_result.group("kconfig_name")
        kconfig_value = re_result.group("kconfig_value")
        if kconfig_value == "y":
            # boolean value
            kconfig_value = True
        elif kconfig_value.startswith("0x") and all(c in string.hexdigits for c in kconfig_value[2:]):
            # hexadecimal value
            kconfig_value = int(kconfig_value, base=16)
        elif kconfig_value.startswith('"') and kconfig_value.endswith('"'):
            # string value
            kconfig_value = kconfig_value[1:-1]
        elif kconfig_value.isdecimal():
            # int value
            kconfig_value = int(kconfig_value, base=10)

        print(f"{kconfig_name=}, {kconfig_value=}")
    else:
        print("No match")

_test_code("CONFIG_TEST=")
_test_code("CONFIG_TEST=\"TEST")
_test_code("CONFIG_TEST=TE\"ST")
_test_code("=")
_test_code("")
_test_code("not_valid_hex=0xWRRRRR")
```
```
config_line='CONFIG_TEST='
kconfig_name='CONFIG_TEST', kconfig_value=''
config_line='CONFIG_TEST="TEST'
kconfig_name='CONFIG_TEST', kconfig_value='"TEST'
config_line='CONFIG_TEST=TE"ST'
kconfig_name='CONFIG_TEST', kconfig_value='TE"ST'
config_line='='
No match
config_line=''
No match
config_line='not_valid_hex=0xWRRRRR'
kconfig_name='not_valid_hex', kconfig_value='0xWRRRRR'
```